### ISSUE TITLE
Zombie service - unable to delete service due to "Unknown service" error

### ISSUE THREAD

**Issue by pySilver** [hashicorp/consul#7669]
#### Overview of the Issue

I'm facing zombie services that are impossible to delete from consul. Zombie appeared after shutting down fabio-lb which does not have a way to remove its service.

```
$ curl \
    --header "X-Consul-Token: ххххххх" \
    http://127.0.0.1:8500/v1/catalog/service/fabio|jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2134    0  2134    0     0   347k      0 --:--:-- --:--:-- --:--:--  416k
[
  {
    "ID": "5d0069e5-0f6e-7d2b-0391-6383646a0edb",
    "Node": "hashistack-1",
    "Address": "10.0.0.3",
    "Datacenter": "dc1",
    "TaggedAddresses": {
      "lan": "10.0.0.3",
      "lan_ipv4": "10.0.0.3",
      "wan": "10.0.0.3",
      "wan_ipv4": "10.0.0.3"
    },
    "NodeMeta": {
      "consul-network-segment": ""
    },
    "ServiceKind": "",
    "ServiceID": "fabio-hashistack-1-9998",
    "ServiceName": "fabio",
    "ServiceTags": [],
    "ServiceAddress": "127.0.0.1",
    "ServiceTaggedAddresses": {
      "lan_ipv4": {
        "Address": "127.0.0.1",
        "Port": 9998
      },
      "wan_ipv4": {
        "Address": "127.0.0.1",
        "Port": 9998
      }
    },
    "ServiceWeights": {
      "Passing": 1,
      "Warning": 1
    },
    "ServiceMeta": {},
    "ServicePort": 9998,
    "ServiceEnableTagOverride": false,
    "ServiceProxy": {
      "MeshGateway": {},
      "Expose": {}
    },
    "ServiceConnect": {},
    "CreateIndex": 364134,
    "ModifyIndex": 364134
  },
  {
    "ID": "ae8b1959-da5a-1f15-7941-3caf933bdf28",
    "Node": "hashistack-2",
    "Address": "10.0.0.2",
    "Datacenter": "dc1",
    "TaggedAddresses": {
      "lan": "10.0.0.2",
      "lan_ipv4": "10.0.0.2",
      "wan": "10.0.0.2",
      "wan_ipv4": "10.0.0.2"
    },
    "NodeMeta": {
      "consul-network-segment": ""
    },
    "ServiceKind": "",
    "ServiceID": "fabio-hashistack-2-9998",
    "ServiceName": "fabio",
    "ServiceTags": [],
    "ServiceAddress": "127.0.0.1",
    "ServiceTaggedAddresses": {
      "lan_ipv4": {
        "Address": "127.0.0.1",
        "Port": 9998
      },
      "wan_ipv4": {
        "Address": "127.0.0.1",
        "Port": 9998
      }
    },
    "ServiceWeights": {
      "Passing": 1,
      "Warning": 1
    },
    "ServiceMeta": {},
    "ServicePort": 9998,
    "ServiceEnableTagOverride": false,
    "ServiceProxy": {
      "MeshGateway": {},
      "Expose": {}
    },
    "ServiceConnect": {},
    "CreateIndex": 364136,
    "ModifyIndex": 364136
  },
  {
    "ID": "3de71c3d-0bab-a0f9-6fc0-70ae9346ed4e",
    "Node": "hashistack-3",
    "Address": "10.0.0.4",
    "Datacenter": "dc1",
    "TaggedAddresses": {
      "lan": "10.0.0.4",
      "lan_ipv4": "10.0.0.4",
      "wan": "10.0.0.4",
      "wan_ipv4": "10.0.0.4"
    },
    "NodeMeta": {
      "consul-network-segment": ""
    },
    "ServiceKind": "",
    "ServiceID": "fabio-hashistack-3-9998",
    "ServiceName": "fabio",
    "ServiceTags": [],
    "ServiceAddress": "127.0.0.1",
    "ServiceTaggedAddresses": {
      "lan_ipv4": {
        "Address": "127.0.0.1",
        "Port": 9998
      },
      "wan_ipv4": {
        "Address": "127.0.0.1",
        "Port": 9998
      }
    },
    "ServiceWeights": {
      "Passing": 1,
      "Warning": 1
    },
    "ServiceMeta": {},
    "ServicePort": 9998,
    "ServiceEnableTagOverride": false,
    "ServiceProxy": {
      "MeshGateway": {},
      "Expose": {}
    },
    "ServiceConnect": {},
    "CreateIndex": 364140,
    "ModifyIndex": 364140
  }
]
```
So it says there are services placed on this nodes (this are master nodes).

Lets take a look at them:

```
$ curl --request GET --header "X-Consul-Token: hidden" http://127.0.0.1:8500/v1/agent/services | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2416    0  2416    0     0   393k      0 --:--:-- --:--:-- --:--:--  393k
{
  "_nomad-server-idnwzuk2vfpiamh2nh4t5ez2j3ya2qo4": {
    "ID": "_nomad-server-idnwzuk2vfpiamh2nh4t5ez2j3ya2qo4",
    "Service": "nomad-servers",
    # skipped ...
  },
  "_nomad-server-kga3nq34kjs2e7gxwp5pjqdfklc4hncn": {
    "ID": "_nomad-server-kga3nq34kjs2e7gxwp5pjqdfklc4hncn",
    "Service": "nomad-servers",
    # skipped ...
  },
  "_nomad-server-xzo5ogzhft2647anjmvn5f53k4c7ltqs": {
    "ID": "_nomad-server-xzo5ogzhft2647anjmvn5f53k4c7ltqs",
    "Service": "nomad-servers",
    # skipped ...
  },
  "consul-ui": {
    "ID": "consul-ui",
    "Service": "consul-ui",
    # skipped ...
  },
  "vault:10.0.0.4:8200": {
    "ID": "vault:10.0.0.4:8200",
    "Service": "vault",
    # skipped ...
  }
}
```

I've repeated it on all 3 nodes from first output and no, they don't have this service.

I've tried to remove them in many ways, but it's not working:

```
$ curl --request PUT --header "X-Consul-Token: xxx" http://127.0.0.1:8500/v1/agent/service/deregister/fabio-hashistack-3-9998
Unknown service {"fabio-hashistack-3-9998" {}}#

$ curl --request PUT --header "X-Consul-Token: xxx" http://127.0.0.1:8500/v1/agent/service/deregister/5d0069e5-0f6e-7d2b-0391-6383646a0edb
Unknown service {"5d0069e5-0f6e-7d2b-0391-6383646a0edb" {}}#


$ consul services deregister -id=fabio-hashistack-3-9998
Error registering service "": Unexpected response code: 500 (Unknown service {"fabio-hashistack-3-9998" {}})


$ consul services deregister -id=ae8b1959-da5a-1f15-7941-3caf933bdf28
Error registering service "": Unexpected response code: 500 (Unknown service {"ae8b1959-da5a-1f15-7941-3caf933bdf28" {}})
```
 

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
$ consul info
agent:
	check_monitors = 0
	check_ttls = 1
	checks = 5
	services = 5
build:
	prerelease =
	revision = 9ea1a204
	version = 1.7.2
consul:
	acl = enabled
	bootstrap = false
	known_datacenters = 1
	leader = false
	leader_addr = 10.0.0.2:8300
	server = true
raft:
	applied_index = 395702
	commit_index = 395702
	fsm_pending = 0
	last_contact = 58.097447ms
	last_log_index = 395702
	last_log_term = 322
	last_snapshot_index = 393334
	last_snapshot_term = 322
	latest_configuration = [{Suffrage:Voter ID:ae8b1959-da5a-1f15-7941-3caf933bdf28 Address:10.0.0.2:8300} {Suffrage:Voter ID:3de71c3d-0bab-a0f9-6fc0-70ae9346ed4e Address:10.0.0.4:8300} {Suffrage:Voter ID:5d0069e5-0f6e-7d2b-0391-6383646a0edb Address:10.0.0.3:8300}]
	latest_configuration_index = 0
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 322
runtime:
	arch = amd64
	cpu_count = 1
	goroutines = 114
	max_procs = 1
	os = linux
	version = go1.13.7
serf_lan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 73
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 209
	members = 13
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = true
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 82
	members = 3
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details

Ubuntu 18.04

**Comment by pySilver** [comment]
Tried https://github.com/myENA/consul-zombie with no luck.

```
# root @ hashistack-3 in ~/workbook/consul-zombie on git:master o [6:54:01] C:1
$ ./zombie -vvv -s fabio search
+--------------+-------------------------+--------------+----------------+--------------+
|     NODE     |           ID            |     NAME     |    ADDRESS     |    STATE     |
+--------------+-------------------------+--------------+----------------+--------------+
| hashistack-2 | fabio-hashistack-2-9998 | fabio        | 127.0.0.1:9998 | healthy=true |
| hashistack-3 | fabio-hashistack-3-9998 | fabio        | 127.0.0.1:9998 | healthy=true |
| hashistack-1 | fabio-hashistack-1-9998 | fabio        | 127.0.0.1:9998 | healthy=true |
+--------------+-------------------------+--------------+----------------+--------------+
|   SUMMARY    |                           UNHEALTHY: 0 |   HEALTHY: 3   |   TOTAL: 3   |
+--------------+-------------------------+--------------+----------------+--------------+

# root @ hashistack-3 in ~/workbook/consul-zombie on git:master o [6:54:15]
$ ./zombie -vvv -f -s fabio kill
Deregistering fabio: fabio-hashistack-2-9998 (10.0.0.2:8500)
2020/04/20 06:54:27 Unable to deregister: Unexpected response code: 500 (Unknown service {"fabio-hashistack-2-9998" {}})
Deregistering fabio: fabio-hashistack-1-9998 (10.0.0.3:8500)
2020/04/20 06:54:27 Unable to deregister: Unexpected response code: 500 (Unknown service {"fabio-hashistack-1-9998" {}})
Deregistering fabio: fabio-hashistack-3-9998 (10.0.0.4:8500)
2020/04/20 06:54:27 Unable to deregister: Unexpected response code: 500 (Unknown service {"fabio-hashistack-3-9998" {}})
```

So strange it says its healthy. Maybe it did not had any checks.

**Comment by pySilver** [comment]
Grepping data directory:

```
# root @ hashistack-1 in ~ [6:30:21]
$ grep -rnw '/var/consul/' -e 'fabio'
Binary file /var/consul/raft/raft.db matches
Binary file /var/consul/raft/snapshots/320-376936-1587298187828/state.bin matches
Binary file /var/consul/raft/snapshots/322-393327-1587357128564/state.bin matches
```

**Comment by pySilver** [comment]
Strange logs:

```
2020-04-20T07:11:08.888Z [ERROR] agent.http: Request error: method=PUT url=/v1/agent/service/deregister/fabio-hashistack-3-9998 from=127.0.0.1:58900 error="Unknown service {"fabio-hashistack-3-9998" {}}"
2020-04-20T07:11:13.428Z [WARN]  agent: Service deregistration blocked by ACLs: service=fabio-hashistack-3-9998 accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:13.431Z [WARN]  agent: Check deregistration blocked by ACLs: check=fabio-hashistack-3-9998-ttl accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:13.433Z [WARN]  agent: Check deregistration blocked by ACLs: check=service:fabio-hashistack-3-9998:2 accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:18.153Z [WARN]  agent: Service deregistration blocked by ACLs: service=fabio-hashistack-3-9998 accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:18.157Z [WARN]  agent: Check deregistration blocked by ACLs: check=fabio-hashistack-3-9998-ttl accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:18.159Z [WARN]  agent: Check deregistration blocked by ACLs: check=service:fabio-hashistack-3-9998:2 accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:22.766Z [WARN]  agent: Service deregistration blocked by ACLs: service=fabio-hashistack-3-9998 accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:22.771Z [WARN]  agent: Check deregistration blocked by ACLs: check=fabio-hashistack-3-9998-ttl accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
2020-04-20T07:11:22.773Z [WARN]  agent: Check deregistration blocked by ACLs: check=service:fabio-hashistack-3-9998:2 accessorID=927906cd-8086-38a1-2fd3-4a47d344c9ad
```

**Comment by pySilver** [comment]
I've had `fabio` ACL profile which was removed before service was removed. Fixed by:

* disable acl
* remove services
* enable acl

Looks like a bug? I was authenticated with master token.

**Comment by tommyalatalo** [comment]
I've had a similar problem with a fabio ACL policy as well, I get the "deregistration blocked by ACL" error, but with a blank accessorID.

**Comment by beeeeeeenny** [comment]
I encountered the same problem, how can I solve it?

**Comment by TheSpbra1n** [comment]
Same problem with Hashicorp Nomad.
Consul v1.7.2

**Comment by pySilver** [comment]
Got hit again by the same issue. Consul became useless once node ip changed. Services simply will not go away aswell as node. Gosh, I wish there was a tool to force service/node/check removal. It is so painful to deal with stale items.

**Comment by pySilver** [comment]
Seeing this with 1.9.2 constantly in logs: "Check deregistration blocked by ACLs"

**Comment by dnephin** [comment]
Thank you for the bug report! Sorry it has taken us some time to respond. I'm reading over this issue and a few that linked to it.

Some context about the problem:
* Consul has two views of services and checks. There is a centralized "catalog" view that is maintained by the Servers, and there is a local "agent" view on each client agent.
* Generally, the centralized catalog view is updated by the individual agents periodically.
* To perform an update of a service (including deleting it) the client agent uses the ACL token that was used to register the service. If no token was used it tries to use the [default token](https://www.consul.io/docs/agent/options#acl_tokens_default) configured for that agent.

I believe the analysis so far is correct. This problem can occur when the ACL token that was used to register the service is deleted before the service itself. 

To fix the problem, I believe you should be able to use the [catalog deregister API](https://www.consul.io/api-docs/catalog#deregister-entity) instead of the agent register API. Once the service is removed from the catalog, the local agents will stop attempting the deletes.

How were the ACL tokens used to register these services created (and deleted)? Is that handled by Nomad and/or fabio?

**Comment by dnephin** [comment]
We now have at least 3 issues about this problem. I believe this was the first report, but there's a little more debugging and context in #9577, so i'm going to close this issue and track the fix in #9577. 

I've posted a similar summary in that issue here: https://github.com/hashicorp/consul/issues/9577#issuecomment-771203024
